rimage/transform/parallel_projection.go

```diff
@@ -95,5 +97,169 @@ func (pp *ParallelProjection) PointCloudToRGBD(cloud pointcloud.PointCloud) (*ri
 func (pp *ParallelProjection) ImagePointTo3DPoint(pt image.Point, d rimage.Depth) (r3.Vector, error) {
-	return r3.Vector{float64(pt.X), float64(pt.Y), float64(d)}, nil
+	return r3.Vector{X: float64(pt.X), Y: float64(pt.Y), Z: float64(d)}, nil
+}
+
+// ParallelProjectionOntoXZWithRobotMarker allows the creation of a 2D projection of a pointcloud and robot
+// position onto the XZ plane.
+type ParallelProjectionOntoXZWithRobotMarker struct {
+	robotPose *spatialmath.Pose
+}
+
+var (
+	sigmaLevel       = 7
+	imageHeight      = 1080
+	imageWidth       = 1080
+	missThreshold    = 0.44
+	hitThreshold     = 0.53
+	voxelSize        = 1
+	robotMarkerRatio = 5
+)
+
+// PointCloudToRGBD creates an image of a pointcloud in the XZ plane, scaling the points to a standard image
+// size. It will also add a red marker to the map to represent the location of the robot.
+func (ppRM *ParallelProjectionOntoXZWithRobotMarker) PointCloudToRGBD(cloud pointcloud.PointCloud,
+) (*rimage.Image, *rimage.DepthMap, error) {
+	meta := cloud.MetaData()
+
+	var X, Z []float64
+	cloud.Iterate(0, 0, func(pt r3.Vector, data pointcloud.Data) bool {
+		X = append(X, pt.X)
+		Z = append(Z, pt.Z)
+		return true
+	})
+
+	// Calculate max and min range to be represented in the output image and, if needed, cropping it based on
+	// the mean and standard deviation of the X and Z coordinates
+	meanX, stdevX, err := calculateMeanAndStandardDeviation(X)
+	if err != nil {
+		return nil, nil, err
+	}
+	meanZ, stdevZ, err := calculateMeanAndStandardDeviation(Z)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	maxX := math.Min(meanX+float64(sigmaLevel)*stdevX, meta.MaxX)
+	minX := math.Max(meanX-float64(sigmaLevel)*stdevX, meta.MinX)
+	maxZ := math.Min(meanZ+float64(sigmaLevel)*stdevZ, meta.MaxZ)
+	minZ := math.Max(meanZ-float64(sigmaLevel)*stdevZ, meta.MinZ)
+
+	// Change the max and min values to ensure the robot marker can be represented in the output image
+	var robotMarker spatialmath.Pose
+	if ppRM.robotPose != nil {
+		robotMarker = *ppRM.robotPose
+		maxX = math.Max(maxX, robotMarker.Point().X)
+		minX = math.Min(minX, robotMarker.Point().X)
+		maxZ = math.Max(maxZ, robotMarker.Point().Z)
+		minZ = math.Min(minZ, robotMarker.Point().Z)
+	}
+
+	// Calculate the scale factors
+	widthScaleFactor := float64(imageWidth) / (maxX - minX)
+	heightScaleFactor := float64(imageHeight) / (maxZ - minZ)
+
+	// Add points in the pointcloud to a new image
+	color := rimage.NewImage(imageWidth, imageHeight)
+	cloud.Iterate(0, 0, func(pt r3.Vector, data pointcloud.Data) bool {
+		j := (pt.X - meta.MinX) * widthScaleFactor
+		i := (pt.Z - meta.MinZ) * heightScaleFactor
+		x, y := int(math.Round(i)), int(math.Round(j))
+
+		// Adds a point to an image using the value to define the color. If no value is available,
+		// the default color of white is used.
+		if x >= 0 && x < imageWidth && y >= 0 && y < imageHeight && data != nil {
```

Comment:
Will `data` be non-nil for pcds from orbslam?

Reply:
No, ReadPCD always returns a non-nil data. It defaults to using NewBasicData():

```
func NewBasicData() Data {
	return &basicData{}
}
```

where the hasColor and hasValue are false. That being said, is there anything we want to do with the potential color from a PCD?

Reply:
I think not for now, since we don't expect color from slam maps. You could consider adding to the comment above the function that color will be ignored.